File: linked_list/js/single-linked-list.ts
interface CustomNode {
  data: any;
  next: CustomNode | null;
}

class LinkedListNode {
  head!: CustomNode;

  constructor(data: any) {
    this.head = { data, next: null };
  }

  insertToHead(data: any) {}

  insertToLast(data: any) {}

  insertByIndex() {}

  updateByIndex() {}

  deleteFirst() {}

  deleteLast() {}

  deleteByValue() {}

  deleteByIndex() {}
}

const main = () => {
  const list1 = new LinkedListNode(10);
};
main();

File: linked_list/cpp/single_linked_list.cpp
#include <iostream>
#include "single_linked_list.h"
using namespace std;

template <class T>
single_linked_list<T>::single_linked_list(T data)
{
    Node<T> *node = new Node<T>(data);
    head = node;
};

template <class T>
bool single_linked_list<T>::insert_to_head(T data)
{
    Node<T> *newHead = new Node<T>(data, head);
    head = newHead;
    return true;
};

template <class T>
void single_linked_list<T>::forEach(void (*callback_ptr)(T data))
{
    Node<T> *tempNode = head;
    while (tempNode != nullptr)
    {
        callback_ptr(tempNode->data);
        tempNode = tempNode->next;
    }
}

template <class T>
void print_list(T data)
{
    cout << "data: " << data << endl;
}

int main()
{
    single_linked_list<int> list1(2);
    list1.insert_to_head(1);
    list1.insert_to_head(0);
    list1.forEach(print_list);
}